src/util/stream_generator.h

```diff
@@ -117,6 +96,73 @@ class stream_generator_promise {
  private:
   bool isBufferLargeEnough() { return _value.length() >= MIN_BUFFER_SIZE; }
 };
+
+struct stream_generator_sentinel {};
+
+template <size_t MIN_BUFFER_SIZE>
+class stream_generator_iterator {
+  using promise_type = stream_generator_promise<MIN_BUFFER_SIZE>;
+  using coroutine_handle = std::coroutine_handle<promise_type>;
+
+ public:
+  using iterator_category = std::input_iterator_tag;
+  // What type should we use for counting elements of a potentially infinite
+  // sequence?
+  using difference_type = std::ptrdiff_t;
+  using value_type = typename promise_type::value_type;
+  using reference = typename promise_type::reference_type;
+  using pointer = typename promise_type::pointer_type;
+
+  // Iterator needs to be default-constructible to satisfy the Range concept.
+  stream_generator_iterator() noexcept : _coroutine(nullptr) {}
+
+  explicit stream_generator_iterator(coroutine_handle coroutine) noexcept
+      : _coroutine(coroutine) {}
+
+  friend bool operator==(const stream_generator_iterator& it,
+                         stream_generator_sentinel) noexcept {
+    return !it._coroutine ||
+           (it._coroutine.done() && it._coroutine.promise().value().empty());
+  }
+
+  friend bool operator!=(const stream_generator_iterator& it,
+                         stream_generator_sentinel s) noexcept {
+    return !(it == s);
+  }
+
+  friend bool operator==(stream_generator_sentinel s,
+                         const stream_generator_iterator& it) noexcept {
+    return (it == s);
+  }
+
+  friend bool operator!=(stream_generator_sentinel s,
+                         const stream_generator_iterator& it) noexcept {
+    return it != s;
+  }
+
+  stream_generator_iterator& operator++() {
+    _coroutine.promise().value().clear();
+    if (!_coroutine.done()) {
+      _coroutine.resume();
+    }
+    if (_coroutine.done()) {
+      _coroutine.promise().rethrow_if_exception();
+    }
+
+    return *this;
+  }
+
+  // Need to provide post-increment operator to implement the 'Range' concept.
```

Comment:
Add `The void return type disables any invalid uses, since this iterator does not support post-increment`